examples/chip-tool/commands/common/Command.cpp

```diff
@@ -0,0 +1,135 @@
+/*
+ *   Copyright (c) 2020 Project CHIP Authors
+ *   All rights reserved.
+ *
+ *   Licensed under the Apache License, Version 2.0 (the "License");
+ *   you may not use this file except in compliance with the License.
+ *   You may obtain a copy of the License at
+ *
+ *       http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *   Unless required by applicable law or agreed to in writing, software
+ *   distributed under the License is distributed on an "AS IS" BASIS,
+ *   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *   See the License for the specific language governing permissions and
+ *   limitations under the License.
+ *
+ */
+
+#include "Command.h"
+#include <sstream>
+#include <support/CHIPLogging.h>
+
+bool Command::InitArguments(int argc, char * argv[])
+{
+    bool isValidCommand = false;
+
+    SuccessOrExit(strcmp(name, argv[0]) != 0);
+    VerifyOrExit(argsCount == (size_t)(argc - 1),
+                 ChipLogProgress(chipTool, "InitArgs: Wrong arguments number: %zu instead of %zu", argc - 1, argsCount));
+
+    for (size_t i = 0; i < argsCount; i++)
+    {
+        SuccessOrExit(!InitArgument(i, argv[1 + i]));
+    }
+
+    isValidCommand = true;
+
+exit:
+    return isValidCommand;
+}
+
+bool Command::InitArgument(size_t argIndex, const char * argValue)
+{
+    bool isValidArgument = false;
+
+    Argument * arg = args[argIndex];
+    switch (arg->type)
+    {
+    case ArgumentType::Attribute: {
+        char * value    = reinterpret_cast<char *>(arg->value);
+        isValidArgument = (strcmp(argValue, value) == 0);
+        break;
+    }
+
+    case ArgumentType::Number: {
+        uint32_t * value = reinterpret_cast<uint32_t *>(arg->value);
+        // stringstream treats uint8_t as char, which is not what we want here.
+        std::stringstream ss(argValue);
+        ss >> *value;
+        isValidArgument = (!ss.fail() && ss.eof() && *value >= arg->min && *value <= arg->max);
+        break;
+    }
+
+    case ArgumentType::Address: {
+        Inet::IPAddress * value = reinterpret_cast<Inet::IPAddress *>(arg->value);
+        isValidArgument         = IPAddress::FromString(argValue, *value);
+        break;
+    }
+    }
+
+    if (!isValidArgument)
+    {
+        ChipLogError(chipTool, "InitArgs: Invalid argument %s: %s", arg->name, argValue);
+    }
+
+    return isValidArgument;
+}
+
+size_t Command::AddArgument(const char * name, const char * value)
+{
+    Argument * arg = new Argument();
```

Comment:
Can we avoid the bare new here? We can use `std::unique_ptr` if the allocation is important here.

Reply:
Done.